### PR TITLE
BaSP-TG-26: Add/Edit Employee

### PR DESCRIPTION
**Pull request by Javier-Cervera**
# Description of functioning --> Not for QA?
### **FormModal** component starts the Add/POST functionality

1. **FormModal** nests, and is the visual structure, for the **Form** component.
2. **Form** contains the logic for creating/updating an Employee. Also, it nests **Input** components for better modularization.
3. When the submit button is pressed a **SuccessModal** component opens, showing the message of success or error. This message comes from a prop from **FormModal**.

### Inside **List** component --> **ListItem** component--> **FormModal** starts the Edit/PUT functionality with a prop `edit={true}` from **ListItem**
-  Also, **ListItem** passes `listItem` prop that holds the Employee object from the DB.

1. **FormModal** passes `listItem` and `edit` props to **Form**.
2. WIth `edit` as a condition, Employee data from the DB is inserted into the **Form** Inputs.
3. With the same condition, PUT request is used.
4. The rest works like in Add/POST.

## Notes

- **List** and childs from [TG-32 List employees](https://github.com/BaSP-m2022/alexis-trackgenix-app/tree/feature-TG-32-list-employees) added to implement Edit functionality --> Might need to merge TG-32 and pull changes so that this PR can be more easily reviewed.
- `.toString().replace(/\s+/g, '').split(',')` in Form is used to standardize inputs sent in either of the requests because sometimes it comes as an array and sometimes as a string, and it needs to be an array. And to delete whitespace.

### CODE REVIEW

File: src/Components/Employees/SuccessModal/index.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import styles from './success-modal.module.css';
+
+const SuccessModal = (props) => {
+  if (!props.show) {
+    return null;
+  }
+  console.log(props.message);
```
Comment (agusting19) [inline src/Components/Employees/SuccessModal/index.js:7]: console.log

File: src/Components/Employees/ListItem/ListItem.jsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    try {
+      await fetch(`${process.env.REACT_APP_API_URL}api/employees/${_id}`, {
+        method: 'DELETE'
+      });
+      setShowModal(true);
+      setTitleModal('Employee deleted');
+    } catch (error) {
+      setShowModal(true);
+      setTitleModal('Error. Employee could not be deleted');
+      console.error(error);
```
Comment (agusting19) [inline src/Components/Employees/ListItem/ListItem.jsx:23]: console.log

**Review by agusting19 — CHANGES_REQUESTED** [review]
The only thing to change is the fetch url and the console log's

File: src/Components/Employees/SuccessModal/index.js (hunk 1 above)
Comment (Javier-Cervera) [reply src/Components/Employees/SuccessModal/index.js]: Removed, thanks!

File: src/Components/Employees/ListItem/ListItem.jsx (hunk 2 above)
Comment (Javier-Cervera) [reply src/Components/Employees/ListItem/ListItem.jsx:23]: Corresponds to TG-32

File: src/Components/Employees/index.js [hunk 3]
```diff
@@ -4,19 +4,20 @@ import List from './List/List';
...
+  const [showFormModal, setShowFormModal] = useState(false);
   const [list, setEmployees] = useState([]);
 
   useEffect(() => {
-    fetch(`${process.env.REACT_APP_API_URL}api/employees/`)
+    fetch(`${process.env.REACT_APP_API_URL}/api/employees/`)
       .then((response) => response.json())
       .then((response) => {
         setEmployees(response.data);
+        console.log(response.data);
```
Comment (agusting19) [inline src/Components/Employees/index.js:15]: console.log

**Review by agusting19 — COMMENTED** [review]
Looks nice, only one missed console.log

File: src/Components/Employees/index.js (hunk 3 above)
Comment (Javier-Cervera) [reply src/Components/Employees/index.js:15]: Thanks!

**Comment by tinchorova** [comment]
Testing Results
Tested locally on https://alexis-trackgenix-app-git-feature-tg-26-add-e-c249ca-basp-m2022.vercel.app/employees

Test Results: Failed ⚠️

Bugs found:

**Add feature**
1. When creating a new employee, an error message pops up saying "active" must be a boolean even though "active" is selected. image
2. The message in 1. is confusing for the user. The user shouldn't know what a boolean is.
3. Same problem going on with  PM boolean.
4. The only way I could create an employee was first selecting the boolean (Active and PM) that was not the default option.


**Edit feature**
6. When filling projects or timesheets with invalid data an error message pops up, this message is not understandable for the user 
image
7. After succesfully edited an employee, it should redirect you to the list of employees, instead it stays in edit window (leaving employee updated message)
image

**PD:** Nice ticket!

**Review by tinchorova — CHANGES_REQUESTED** [review]
Please refer to: https://github.com/BaSP-m2022/alexis-trackgenix-app/pull/2#issuecomment-1145917899

**Review by tinchorova — APPROVED** [review]
Testing Results
Tested locally on https://alexis-trackgenix-app-git-feature-tg-26-add-e-c249ca-basp-m2022.vercel.app/employees

Test Results: approved!

⚠️ KEEP IN MIND: When opening edit or add modals, user shouldn't be allowed to scroll the page ⚠️

**Pull request merged**